pkg/compiler/debug.go

```diff
@@ -0,0 +1,186 @@
+package compiler
+
+import (
+	"encoding/json"
+	"errors"
+	"go/ast"
+	"go/types"
+	"strconv"
+	"strings"
+)
+
+// DebugInfo represents smart-contract debug information.
+type DebugInfo struct {
+	EntryPoint string            `json:"entrypoint"`
+	Documents  []string          `json:"documents"`
+	Methods    []MethodDebugInfo `json:"methods"`
+	Events     []EventDebugInfo  `json:"events"`
+}
+
+// MethodDebugInfo represents smart-contract's method debug information.
+type MethodDebugInfo struct {
+	ID string `json:"id"`
+	// Name is the name of the method in a format "{namespace},{display-name}"
+	Name string `json:"name"`
+	// Range is the range of smart-contract's opcodes corresponding to the method.
+	Range debugRange `json:"range"`
+	// Parameters is a list of method's parameters.
+	Parameters []debugParam `json:"params"`
+	// ReturnType is method's return type.
+	ReturnType string   `json:"return-type"`
+	Variables  []string `json:"variables"`
+	// SeqPoints is a map between source lines and byte-code instruction offsets.
+	// It has following format: "{address}[{document-index}]{start-line}:{start-column}-{end-line}:{end-column}".
+	SeqPoints []string `json:"sequence-points"`
+}
+
+// EventDebugInfo represents smart-contract's event debug information.
+type EventDebugInfo struct {
+	ID string `json:"id"`
+	// Name is a human-readable event name in a format "{namespace}-{name}".
+	Name       string       `json:"name"`
+	Parameters []debugParam `json:"parameters"`
+}
+
+type debugRange struct {
+	Start uint16
+	End   uint16
+}
+
+type debugParam struct {
+	Name string
+	Type string
+}
+
+func (c *codegen) emitDebugInfo() *DebugInfo {
+	d := &DebugInfo{
+		EntryPoint: mainIdent,
+		Events:     []EventDebugInfo{},
+	}
+	for name, scope := range c.funcs {
+		m := c.methodInfoFromScope(name, scope)
+		if m.Range.Start == m.Range.End {
+			continue
+		}
+		d.Methods = append(d.Methods, *m)
+	}
+	return d
+}
+
+func (c *codegen) methodInfoFromScope(name string, scope *funcScope) *MethodDebugInfo {
+	ps := scope.decl.Type.Params
+	params := make([]debugParam, 0, ps.NumFields())
+	for i := range params {
+		for j := range ps.List[i].Names {
+			params = append(params, debugParam{
+				Name: ps.List[i].Names[j].Name,
+				Type: c.scTypeFromExpr(ps.List[i].Type),
+			})
+		}
+	}
+	return &MethodDebugInfo{
+		ID:         name,
+		Name:       "," + name,
+		Range:      scope.rng,
+		Parameters: params,
+		ReturnType: c.scReturnTypeFromScope(scope),
+	}
+}
+
+func (c *codegen) scReturnTypeFromScope(scope *funcScope) string {
+	results := scope.decl.Type.Results
+	switch results.NumFields() {
+	case 0:
+		return "Void"
+	case 1:
+		return c.scTypeFromExpr(results.List[0].Type)
+	default:
+		// multiple return values are not supported in debugger
```

Comment:
Have you submitted an issue for it?

Reply:
https://github.com/neo-project/neo-debugger/issues/17